src/mpid/ch4/shm/ipc/src/ipc_send.h

```diff
@@ -31,7 +31,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_IPCI_try_lmt_isend(const void *buf, MPI_Aint
 
     MPIDI_POSIX_THREAD_CS_ENTER_VCI(vsi_src);
 
-    if (rank == comm->rank) {
```

Comment:
There shouldn't be any need to use the IPC path when sending to the same rank in a non-self communicator, since the memory doesn't need to be mapped. But I think what you are saying is that the fallback path does an extra copy (into genq buffers) as opposed to just copying from send to receive buffer with MPIR_Localcopy? If so, that does seem like something in need of fixing.

Reply:
Yes, I agree the best way to handle the case of sending data to the same rank in a non-self communicator is to use MPIR_Localcopy.
With this PR, the data transfer between the same rank will use IPC P2P, which is not optimal, but it is much better than using POSIX P2P.